crossopt/assignment4/countislands.py

```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env python3
+
+
+"""grid of tiles. stores islands via dsu"""
+"""https://en.wikipedia.org/wiki/Disjoint-set_data_structure"""
+class Grid(object):
+    def __init__(self, n, m, grid):
+        self.n = n
+        self.m = m
+        self.grid = grid
+        self.parents = dict()
+        self.sizes = dict()
+        for i in range(n):
+            for j in range(m):
+                self.parents[i, j] = (i, j)
+                self.sizes[i, j] = 1
+                if i and self.grid[i][j] and self.grid[i - 1][j]:
+                    self.join((i, j), ((i - 1, j)))
+                if j and self.grid[i][j] and self.grid[i][j - 1]:
+                    self.join((i, j), ((i, j - 1)))
+
+    # returns unique id for each island
+    def __getitem__(self, u):
```

Comment:
I do not understand what is the role of the getitem. Can you document it?
I found it surprising that a call to __getitem__ mutates the internal state.

Reply:
Specifically, I do not understand when this important step is designed to happen: 
1. at construction time? 
2. or at countIsland time?

As I read it, the mutation can happen at both times.

Reply:
Oh, I now see that __getitem__ is recursive, on top of mutating. This confuses the complexity. 

1. Can you rename __getitem__ to something more meaningful?
2. Can you make sure this method is called only at __init__ or at countIsland time, but not both?

Reply:
getitem returns a single unique item for all of the elements in a set.
The mutation of the parents is just an internal part of the operation that improves the complexity